src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs

```diff
@@ -680,32 +685,80 @@ public string[] GetAvailableProjects()
 
             return NuGetUIThreadHelper.JoinableTaskFactory.Run(async delegate
                 {
-                    await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                    var allProjects = await _solutionManager.GetNuGetProjectsAsync();
+
+                    var projectSafeNames = new List<string>();
+                    var displayNames = new List<string>();
+
+                    var safeNametasks = new List<Task<string>>();
+                    var displayNameTasks = new List<Task<string>>();
+
+                    foreach (var project in allProjects)
+                    {
+                        // Throttle and wait for a task to finish if we have hit the limit
+                        if (safeNametasks.Count == MaxTasks)
+                        {
+                            var safeName = await CompleteTaskAsync(safeNametasks);
+                            projectSafeNames.Add(safeName);
+                        }
+
+                        if (displayNameTasks.Count == MaxTasks)
+                        {
+                            var displayName = await CompleteTaskAsync(displayNameTasks);
+                            displayNames.Add(displayName);
+                        }
 
-                    var allProjects = _solutionManager.GetNuGetProjects();
-                    _projectSafeNames = allProjects.Select(_solutionManager.GetNuGetProjectSafeName).ToArray();
-                    var displayNames = allProjects.Select(GetDisplayName).ToArray();
-                    Array.Sort(displayNames, _projectSafeNames, StringComparer.CurrentCultureIgnoreCase);
+                        var safeNameTask = Task.Run(async () => await _solutionManager.GetNuGetProjectSafeNameAsync(project));
```

Comment:
What's the purpose of all these tasks manipulation? Is this needed for a perf improvement?
Not sure whether this logic is highly efficient. All the generated tasks are UI thread bound so they won't run in parallel but rather will block on each other.